src/main/scala/beam/router/BeamSkimmer.scala

```diff
@@ -390,95 +390,97 @@ class BeamSkimmer @Inject()(val beamConfig: BeamConfig, val beamServices: BeamSe
+  //    writer.write(Eol)
+  //
+  //    beamServices.tazTreeMap.getTAZs
+  //      .foreach { origin =>
+  //        beamServices.tazTreeMap.getTAZs.foreach { destination =>
+  //          uniqueModes.foreach { mode =>
+  //            uniqueTimeBins
+  //              .foreach { timeBin =>
+  //                val theSkim: Skim = getSkimValue(timeBin * 3600, mode, origin.tazId, destination.tazId)
+  //                  .map(_.toSkimExternal)
+  //                  .getOrElse {
+  //                    if (origin.equals(destination)) {
+  //                      val newDestCoord = new Coord(
+  //                        origin.coord.getX,
+  //                        origin.coord.getY + Math.sqrt(origin.areaInSquareMeters) / 2.0
+  //                      )
+  //                      getSkimDefaultValue(
+  //                        mode,
+  //                        origin.coord,
+  //                        newDestCoord,
+  //                        timeBin * 3600,
+  //                        dummyId,
+  //                        beamServices
+  //                      )
+  //                    } else {
+  //                      getSkimDefaultValue(
+  //                        mode,
+  //                        origin.coord,
+  //                        destination.coord,
+  //                        timeBin * 3600,
+  //                        dummyId,
+  //                        beamServices
+  //                      )
+  //                    }
+  //                  }
+  //
+  //                writer.write(
+  //                  s"$timeBin,$mode,${origin.tazId},${destination.tazId},${theSkim.time},${theSkim.generalizedTime},${theSkim.cost},${theSkim.generalizedTime},${theSkim.distance},${theSkim.count},${theSkim.energy}\n"
+  //                )
+  //              }
+  //          }
+  //        }
+  //      }
+  //    writer.close()
+  //  }
 
   def writeObservedSkims(event: IterationEndsEvent): Unit = {
-    val fileHeader =
-      "hour,mode,origTaz,destTaz,travelTimeInS,generalizedTimeInS,cost,generalizedCost,distanceInM,numObservations,energy"
     val filePath = event.getServices.getControlerIO.getIterationFilename(
       event.getServices.getIterationNumber,
       BeamSkimmer.observedSkimsFileBaseName + ".csv.gz"
     )
+    //    val writer = IOUtils.getBufferedWriter(filePath)
     val writer = IOUtils.getBufferedWriter(filePath)
-    writer.write(fileHeader)
-    writer.write("\n")
-
-    skims.foreach { keyVal =>
-      writer.write(
-        s"${keyVal._1._1},${keyVal._1._2},${keyVal._1._3},${keyVal._1._4},${keyVal._2.time},${keyVal._2.generalizedTime},${keyVal._2.cost},${keyVal._2.generalizedCost},${keyVal._2.distance},${keyVal._2.count},${keyVal._2.energy}\n"
-      )
+    try {
+      toCsv(skims).foreach(writer.write)
+    } finally {
+      writer.close()
     }
-    writer.close()
   }
+
 }
 
 object BeamSkimmer extends LazyLogging {
+  type BeamSkimmerKey = (Int, BeamMode, Id[TAZ], Id[TAZ])
```

Comment:
I know you just refactored, but what about extracting it as case class instead of Tuple4?

Reply:
I like the idea, but don want to change too much in this PR since its purpose was only to create a test. Small refactoring was only to avoid duplication code